include/deal.II/base/mpi_remote_point_evaluation.h

```diff
@@ -448,8 +479,189 @@ namespace Utilities
+       * - point data (sorted according to ranks) -> to be sent
+       * - memory for receiving data (by one process at a time)
+       *
+       */
+      unsigned int buffer_size_with_sorting;
+
+      /**
+       * Buffer size (if sorting is not requested); corresponds to the
+       * number of evaluation points.
+       */
+      unsigned int buffer_size_without_sorting;
     };
 
+    namespace internal
+    {
+#ifdef DEAL_II_WITH_MPI
+      /**
+       * Pack @p data and send it via MPI_Isend.
+       */
+      template <typename T>
+      std::enable_if_t<Utilities::MPI::is_mpi_type<T> == false, void>
+      pack_and_isend(const ArrayView<const T>       &data,
+                     const unsigned int              rank,
+                     const unsigned int              tag,
+                     const MPI_Comm                  comm,
+                     std::vector<std::vector<char>> &buffers,
+                     std::vector<MPI_Request>       &requests)
+      {
+        requests.emplace_back(MPI_Request());
+
+        buffers.emplace_back(Utilities::pack(
+          std::vector<T>(data.data(), data.data() + data.size()), false));
+
+        const int ierr = MPI_Isend(buffers.back().data(),
+                                   buffers.back().size(),
+                                   MPI_CHAR,
+                                   rank,
+                                   tag,
+                                   comm,
+                                   &requests.back());
+        AssertThrowMPI(ierr);
+      }
+
+
+
+      /**
+       * Above function specialized for data types supported by MPI
+       * so that one can skip packing.
+       */
+      template <typename T>
+      std::enable_if_t<Utilities::MPI::is_mpi_type<T> == true, void>
+      pack_and_isend(const ArrayView<const T>       &data,
+                     const unsigned int              rank,
+                     const unsigned int              tag,
+                     const MPI_Comm                  comm,
+                     std::vector<std::vector<char>> &buffers,
+                     std::vector<MPI_Request>       &requests)
+      {
+        (void)buffers;
+
+        requests.emplace_back(MPI_Request());
+
+        const int ierr = MPI_Isend(data.data(),
+                                   data.size(),
+                                   Utilities::MPI::mpi_type_id_for_type<T>,
+                                   rank,
+                                   tag,
+                                   comm,
+                                   &requests.back());
+        AssertThrowMPI(ierr);
+      }
+
+
+
+      /**
+       * Above function specialized for Tenors. The underlying data type might
+       * be supported by MPI so that one can skip packing.
+       */
+      template <int rank_, int dim, typename T>
+      std::enable_if_t<Utilities::MPI::is_mpi_type<T> == true, void>
+      pack_and_isend(const ArrayView<const Tensor<rank_, dim, T>> &data,
```

Comment:
Are you actually using this function in the respective application code? I see that the multigrid application unrolls the data explicitly. I realize that the new interface gives the user the ability call the function both with scalars and `Tensor` arguments, but then I wonder to which extent the `n_components` interface further down is needed, and why we won't just always use the `n_components = 1` case. Could you please explain (and add comments in case my question makes sense)?

Reply:
In the multigrid case, I am unrolling the tensors to be able to store the buffers in the class (independent of the number of dimension and rank). The alternative would be to have multiple buffers with the right type (this is what had sometime during the implementation), which is not salable in particular if we want to support more components. At the moment we only support `n_components` (-> Poisson) and `n_components==dim` (-> elasticity). But for that we will probably need to restructure the code more and the conversion of the number of components to a templated variable much more inner.